src/extensions/Wyam.YouTube/YouTube.cs

```diff
@@ -0,0 +1,113 @@
+    /// </remarks>
+    /// <category>Metadata</category>
+    public class YouTube : IModule, IAsNewDocuments
+    {
+        private readonly BaseClientService.Initializer _initializer;
+        private readonly Dictionary<string, Func<IDocument, IExecutionContext, YouTubeService, object>> _requests
+            = new Dictionary<string, Func<IDocument, IExecutionContext, YouTubeService, object>>();
+
+        /// <summary>
+        /// Creates a connection to the YouTube API with authenticated access.
+        /// </summary>
+        /// <param name="apiKey">The apikey to use.</param>
+        public YouTube(string apiKey)
+        {
+            _initializer = new BaseClientService.Initializer()
+            {
+                ApplicationName = "Wyam",
+                ApiKey = apiKey
+            };
+        }
+
+        /// <summary>
+        /// Submits a request to the YouTube client. This allows you to incorporate data from the execution context in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public YouTube WithRequest(string key, Func<IExecutionContext, YouTubeService, object> request)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentException("Argument is null or empty", nameof(key));
+            }
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            _requests[key] = (doc, ctx, github) => request(ctx, github);
+            return this;
+        }
+
+        /// <summary>
+        /// Submits a request to the YouTube client. This allows you to incorporate data from the execution context and current document in your request.
+        /// </summary>
+        /// <param name="key">The metadata key in which to store the return value of the request function.</param>
+        /// <param name="request">A function with the request to make.</param>
+        /// <returns>The current module instance.</returns>
+        public YouTube WithRequest(string key, Func<IDocument, IExecutionContext, YouTubeService, object> request)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentException("Argument is null or empty", nameof(key));
+            }
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            _requests[key] = request;
+            return this;
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<IDocument> Execute(IReadOnlyList<IDocument> inputs, IExecutionContext context)
+        {
+            // YouTubeService is IDisposable.... how do you want this handled?
```

Comment:
https://github.com/google/google-api-dotnet-client/blob/master/Src/Support/Google.Apis/Services/BaseClientService.cs#L134

https://github.com/google/google-api-dotnet-client/blob/master/Src/Support/Google.Apis/Services/IClientService.cs#L35

Reply:
So here's what I ended up doing - in the most recent commit to develop, any module that implements `IDisposable` will get disposed when the engine is disposed (and recursively through child modules if the module also implements `IEnumerable<IModule>`). That way you can instantiate a long-running service like this in the constructor and then clean it up when the engine is totally finished.